### ISSUE TITLE
[Docs] 4.0.0 parameters missing from model.txt

### ISSUE THREAD

**Issue by adfea9c0** [microsoft/LightGBM#6010]
## Description
I noticed when you save a model, some of the new parameters in the latest 4.0.0 release are not stored, e.g. the ones around quantized training. It seems they are explicitly marked as such in the document:

https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/include/LightGBM/config.h#L596-L622

Is this intentional? It seems to me these parameters ought to be saved.

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

> It seems to me these parameters ought to be saved.

I believe this was intentional in #5800. Those parameters don't affect serving a model to generate predictions (similar conversation: https://github.com/microsoft/LightGBM/pull/6017#discussion_r1301883859).

Could you explain specifically why you think they should be saved?

**Comment by jameslamb** [comment]
I thought about it some more tonight and I support adding these to model files, let's see if other maintainers agree: https://github.com/microsoft/LightGBM/pull/6017#discussion_r1303790867

But either way, please share as much detail as you can about why you think they "ought to be saved".

**Comment by jameslamb** [comment]
By the way, I modified the code link in your initial post to one that's tagged to a specific commit. "lines 596-622 of `config.h`" will point to different code in the future as that file changes, which might make it difficult to understand the intent of this issue. Doing that also has the side benefit of getting a nice inline view of the exact lines you were referring to.

If you're not familiar with that, see https://docs.github.com/en/repositories/working-with-files/using-files/getting-permanent-links-to-files#press-y-to-permalink-to-a-file-in-a-specific-commit.

**Comment by mjmckp** [comment]
Please also ensure `data_sample_strategy` is also persisted to the model file.

**Comment by adfea9c0** [comment]
> But either way, please share as much detail as you can about why you think they "ought to be saved".

Reproducibility and hyperparamger logging in machine learning in general is valuable -- if I go back and look at a `model.txt` I trained earlier, being able to see what hyperparameters I used to train it seems like a nice feature to have. I understand there's limitations to this, we can't put the whole dataset in the model.txt. But a few hyperparameters seem ~free to add.

> those parameters don't affect serving a model to generate predictions (similar conversation: https://github.com/microsoft/LightGBM/pull/6017#discussion_r1301883859).

I think this is true for a lot of parameters that currently _are_ printed? I don't need to know the bagging fraction used for training if I want to do inference. (To be clear, I'm not arguing it should be removed, quite the opposite :-).)

If the concern here is file size, maybe there could be a succint mode that only stores the bare minimum relevant for serving (so no irrelevant hyperparameters, feature importances, tree descriptions only contain splits/values, not counts/weights).

**Comment by jameslamb** [comment]
Thanks for that.

I intentionally wanted to ask in an open-ended, non-leading way to see if there were other specific use cases you were thinking of that I hadn't considered.

Seems like there are not and your motivation for this request is "seems like it'd be nice". To be clear... I don't disagree! We'll pick this up at some point, thanks for taking the time to start a thread about it here.

**Comment by mjmckp** [comment]
LightGBM.net relies on all parameters being present in the model file in
order to validate that the parameters were correctly specified at the input
to training.

**Comment by adfea9c0** [comment]
Also in 4.0.0 I noticed that at some point the python code will actually try to parse parameters from the model text file, namely https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/python-package/lightgbm/basic.py#L3234-L3254

Though in practice the goal seems to be to override input parameters, see 
https://github.com/microsoft/LightGBM/blob/858eeb54a215ff50fd3ca0033b77d51bf48f2c1f/python-package/lightgbm/basic.py#L3184-L3186

So this in practice will end up loading the wrong parameters, e.g.:

```python
import numpy as np
import lightgbm as lgb
assert lgb.__version__ == "4.0.0"

n, p = 10_000, 10
X = np.random.normal(size=(n, p))
y = np.random.normal(size=(n,))
ds = lgb.Dataset(X, label=y, free_raw_data=False)

model = lgb.train({"verbosity": -1, "use_quantized_grad": True}, train_set=ds, num_boost_round=1)
print("use_quantized_grad" in model.params)
>>> True

model.save_model("tmp.txt")

booster = lgb.Booster({"verbosity": -1, "use_quantized_grad": True}, model_file="tmp.txt")
print("use_quantized_grad" in booster.params)
>>> False
```

TBH I don't really know what the idea is behind this parameter override when loading a model, but either way this doesn't really seem correct?

**Comment by jameslamb** [comment]
> was the original reason for omitting these output parameters indeed to keep the output file small for inference?

Not necessarily. The idea of excluding parameters from the model file intentionally originally came from a desire to not include CLI-only parameters to the model file.

For example, the CLI supports parameters like these:

* `output_model` = filepath, if provided tells LightGBM to write a model in text format to that path during training
* `save_binary` = boolean, tells LightGBM to write the training `Dataset` out to LightGBM's binary format

Having such parameters in the model config by default is problematic outside of the CLI, like in the R and Python packages, because it can result in side-effects like file-writing that users didn't ask for and which happen silently.

See this thread: https://github.com/microsoft/LightGBM/pull/2589#discussion_r349897409

(which I found from the git blame: https://github.com/microsoft/LightGBM/blame/82033064005d07ae98cd3003190d675992061a61/include/LightGBM/config.h#L9)

I suspect that other parameters were then later marked `[no-save]` (omitted from the model flle) for other reasons, like maybe a misunderstanding that the model file was supposed to only contain information necessary for inference + whatever it already contained. I'm not sure.

> Would having a smaller inference-only text file format resolve this?

I don't support adding this at this time. The size of LightGBM text files is already dominated by the tree structure (which would be necessary for inference too), so I don't think a code path generating a model file which omits parameters and other model information just for the sake of file size is worth the added complexity.

**Comment by jameslamb** [comment]
> TBH I don't really know what the idea is behind this parameter override when loading a model, but either way this doesn't really seem correct?

I don't understand this comment. What *specifically* is not "correct" about that behavior you linked from the Python package?

**Comment by jameslamb** [comment]
> See the code snippet I posted. I set use_quantized_grad=True in both boosters but it is not present in the parameters in the second booster.

I understand the snippet, and understand that you mean it to say that you *expected* parameters passed to `lgb.train()` to override those loaded from the model.

But the fact that that's what you expected, or that that's what you'd prefer, doesn't mean LightGBM's current approach is not "correct".

That behavior was an intentional choice. If you look in the git blame view for the code you linked to ([blame link](https://github.com/microsoft/LightGBM/blame/1d7ee63686272bceffd522284127573b511df6be/python-package/lightgbm/basic.py#L3242-L3244)), it'll lead you to the PR that introduced it: #5424.

From there, you can find that there was actually quite a lot of conversation about this topic on the PR:

* https://github.com/microsoft/LightGBM/pull/5424#discussion_r959949405
* https://github.com/microsoft/LightGBM/pull/5424#discussion_r991662768

-----

@jmoralez when you have time could you read through those two threads again and give us your opinion on whether parameters passed through `params` in `lgb.train()` should override what's lloaded from a model file?

Re-reading them and considering the conversation in this issue, I think we should have params specified in `lgb.train()` override any loaded from the model file if you pass in a model file.

**Comment by jmoralez** [comment]
The example is constructing a booster passing both params and model_file, which I'm pretty sure issues a warning about the params being ignored. I believe the conclusion was that if the user wanted to train more iterations it'd have to be done with train and passing the saved model as init_model.

**Comment by jameslamb** [comment]
Oh!! I totally missed that it was a call to `lgb.Booster()` there, not a second call to `lgb.train()`. 😅

**Comment by adfea9c0** [comment]
Note that I set `use_quantized_grad=True` for both models. The point is that there is something off, whether my expectations are correct or not. Either the parameters from the `model_file` are leading (the current situation) or the passed in parameters are leading (my expectation), but either way this should lead to `use_quantized_grad=True` in the final model, because I set it to `True` in both dictionaries.

I think there are issues with this setup in `lgb.train` as well. I experimented a bit and the siutation is actually even stranger because whether or not the parameter gets correctly loaded seems to depend on what you pass into the second `lgb.train`. Let's ignore my expectations now -- **I accept that the parameters from the loaded file are leading in the current version**. Consider this snippet:

```python
import numpy as np
import lightgbm as lgb
assert lgb.__version__ == "4.0.0"

np.random.seed(0)
n, p = 10_000, 10
X, y = np.random.normal(size=(n, p)), np.random.normal(size=(n,))
ds = lgb.Dataset(X, label=y, free_raw_data=False)

model = lgb.train({"verbosity": -1, "use_quantized_grad": True}, train_set=ds, num_boost_round=1)
model.save_model("tmp.txt")

model2 = lgb.train({"verbosity": -1}, train_set=ds, init_model="tmp.txt")
print("use_quantized_grad" in model2.params)
>>> False

model3 = lgb.train({"verbosity": -1, "use_quantized_grad": True}, train_set=ds, init_model="tmp.txt")
print("use_quantized_grad" in model3.params)
>>> True

model4 = lgb.train({"verbosity": -1, "use_quantized_grad": False}, train_set=ds, init_model="tmp.txt")
print("use_quantized_grad" in model4.params)
>>> True
```
What is happening is:
 1. I first store a model in `tmp.txt` with `use_quantized_grad=True`
 2. When I use it as an `init_model` for `model2`, this will have `use_quantized_grad=False` (which is against the suggested behaviour which says the parameters from the model file are supposed to be leading).
 3. When I use it as an `init_model` for `model3` and pass in `use_quantized_grad=True`, this will have `use_quantized_grad=True`.
 4.  3. When I use it as an `init_model` for `model3` and pass in `use_quantized_grad=False`, this will have `use_quantized_grad=True` as well!!

I can't really think of a consistent ruleset that explains the above. Either the parameters from the loaded booster are supposed to be leading, in which case `model2` is incorrect. Or the parameters from the input dictionary are supposed to be leading in which case `model4` is incorrect.

**Comment by adfea9c0** [comment]
Just to drive the point home. Can you please have a look at this code snippet and guess what it should output, and then try and run it:
```python
import numpy as np
import lightgbm as lgb
assert lgb.__version__ == "4.0.0"

np.random.seed(0)
n, p = 10_000, 10
X, y = np.random.normal(size=(n, p)), np.random.normal(size=(n,))
ds = lgb.Dataset(X, label=y, free_raw_data=False)

params1 = {"verbosity": -1, "use_quantized_grad": True, "bagging_fraction": 0.5}
params2 = {"verbosity": -1, "use_quantized_grad": False, "bagging_fraction": 1.0}

model = lgb.train(params1, train_set=ds, num_boost_round=1)
model.save_model("tmp.txt")

model2 = lgb.Booster(params2, model_file="tmp.txt")

print(f"{model2.params.get('use_quantized_grad', None)}")
print(f"{model2.params.get('bagging_fraction', None)}")
```

**Comment by jmoralez** [comment]
> I train and save a booster with use_quantized_grad=True (emphasis on True)

This param isn't currently saved in the model file and thus can't be recovered, but it's being addressed in #6077.

> I create a second booster which loads the above booster and also separately inputs use_quantized_grad=True (emphasis on True)

This issues the warning about the params argument being ignored, so it doesn't have any effect.

> The resulting second booster does not have use_quantized_grad set at all, despite me setting it to True twice.

This is because of the two points above.

> Actually, this may be a bug in itself and only happen for use_quantized_grad and other parameters not currently stored in the model file

Yes, since it's not currently saved in the model file there's no way to recover it, however you can override the parameters for continued training with the lgb.train function.

> For lgb.Booster, the claimed behaviour is that the parameters from the model_file are loaded and override those input the booster

This isn't stated anywhere, those parameters are meant to be used on the Python side to get information on the trained model, such as using the learned categorical features (#5246).

**Comment by jameslamb** [comment]
#6077 was just merged, so as of latest `master` and the next LightGBM release, all of the parameters related to quantized training will be stored in the model file.
